File: pixie/dataset/store.py
"""DatasetStore — JSON-file-backed CRUD for Dataset objects."""

from __future__ import annotations

import builtins
import json
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from pixie.config import get_config
from pixie.dataset.models import Dataset
from pixie.storage.evaluable import Evaluable


def _slugify(name: str) -> str:
    """Convert a dataset name to a filesystem-safe slug.

    Lowercase, replace non-alphanumeric runs with ``-``,
    strip leading/trailing ``-``.
    """
    slug = re.sub(r"[^a-z0-9]+", "-", name.lower()).strip("-")
    if not slug:
        raise ValueError(f"Cannot slugify empty or non-alphanumeric name: {name!r}")
    return slug


class DatasetStore:
    """JSON-file-backed CRUD for ``Dataset`` objects.

    Each dataset is stored as ``<dataset_dir>/<slug>.json``.
    The directory is created on first write if it does not exist.

    Args:
        dataset_dir: Override directory. When ``None``, reads from
            ``PixieConfig.dataset_dir`` (env var ``PIXIE_DATASET_DIR``).
    """

    def __init__(self, dataset_dir: str | Path | None = None) -> None:
        if dataset_dir is not None:
            self._dir = Path(dataset_dir)
        else:
            self._dir = Path(get_config().dataset_dir)

    # -- helpers ----------------------------------------------------------

    def _path_for(self, name: str) -> Path:
        return self._dir / f"{_slugify(name)}.json"

    def _ensure_dir(self) -> None:
        self._dir.mkdir(parents=True, exist_ok=True)

    # -- CRUD -------------------------------------------------------------

    def create(self, name: str, items: list[Evaluable] | None = None) -> Dataset:
        """Create a new dataset.

        Args:
            name: Unique dataset name.
            items: Initial evaluable items (default empty).

        Returns:
            The created ``Dataset``.

        Raises:
            FileExistsError: If a dataset with *name* already exists.
        """
        path = self._path_for(name)
        if path.exists():
            raise FileExistsError(f"Dataset already exists: {name!r}")
        dataset = Dataset(name=name, items=tuple(items) if items else ())
        self._write(path, dataset)
        return dataset

    def get(self, name: str) -> Dataset:
        """Load a dataset by name.

        Args:
            name: The dataset name.

        Returns:
            The loaded ``Dataset``.

        Raises:
            FileNotFoundError: If no dataset with *name* exists.
        """
        path = self._path_for(name)
        if not path.exists():
            raise FileNotFoundError(f"Dataset not found: {name!r}")
        return self._read(path)

    def list(self) -> list[str]:
        """Return the names of all stored datasets.

        Reads the ``name`` field from each JSON file in the dataset
        directory. Returns an empty list if the directory does not exist.
        """
        if not self._dir.exists():
            return []
        names: list[str] = []
        for p in sorted(self._dir.glob("*.json")):
            try:
                ds = self._read(p)
                names.append(ds.name)
            except Exception:
                continue  # skip malformed files
        return names

    def list_details(self) -> builtins.list[dict[str, Any]]:
        """Return metadata for every stored dataset.

        Each returned dict contains:

        - ``name``: dataset name
        - ``row_count``: number of evaluable items
        - ``created_at``: file creation timestamp (ISO 8601, UTC)
        - ``updated_at``: file last-modified timestamp (ISO 8601, UTC)

        Returns an empty list if the directory does not exist.
        """
        if not self._dir.exists():
            return []
        rows: list[dict[str, Any]] = []
        for p in sorted(self._dir.glob("*.json")):
            try:
                ds = self._read(p)
                stat = p.stat()
                rows.append(
                    {
                        "name": ds.name,
                        "row_count": len(ds.items),
                        "created_at": _timestamp_to_iso(stat.st_ctime),
                        "updated_at": _timestamp_to_iso(stat.st_mtime),
                    }
                )
            except Exception:
                continue  # skip malformed files
        return rows

    def delete(self, name: str) -> None:
        """Delete a dataset by name.

        Args:
            name: The dataset name.

        Raises:
            FileNotFoundError: If no dataset with *name* exists.
        """
        path = self._path_for(name)
        if not path.exists():
            raise FileNotFoundError(f"Dataset not found: {name!r}")
        path.unlink()

    def append(self, name: str, *items: Evaluable) -> Dataset:
        """Append items to an existing dataset.

        Args:
            name: The dataset name.
            *items: One or more ``Evaluable`` instances to add.

        Returns:
            The updated ``Dataset``.

        Raises:
            FileNotFoundError: If no dataset with *name* exists.
        """
        dataset = self.get(name)
        updated = Dataset(name=dataset.name, items=dataset.items + tuple(items))
        self._write(self._path_for(name), updated)
        return updated

    def remove(self, name: str, index: int) -> Dataset:
        """Remove an item by index from an existing dataset.

        Args:
            name: The dataset name.
            index: Zero-based index of the item to remove.

        Returns:
            The updated ``Dataset``.

        Raises:
            FileNotFoundError: If no dataset with *name* exists.
            IndexError: If *index* is out of range.
        """
        dataset = self.get(name)
        items = list(dataset.items)
        if index < 0 or index >= len(items):
            raise IndexError(
                f"Index {index} out of range for dataset {name!r} with {len(items)} items"
            )
        items.pop(index)
        updated = Dataset(name=dataset.name, items=tuple(items))
        self._write(self._path_for(name), updated)
        return updated

    # -- I/O helpers ------------------------------------------------------

    def _write(self, path: Path, dataset: Dataset) -> None:
        self._ensure_dir()
        data = dataset.model_dump(mode="json")
        path.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")

    def _read(self, path: Path) -> Dataset:
        raw = json.loads(path.read_text(encoding="utf-8"))
        return Dataset.model_validate(raw)


def _timestamp_to_iso(ts: float) -> str:
    """Convert a POSIX timestamp to an ISO 8601 string (UTC)."""
    return datetime.fromtimestamp(ts, tz=timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

File: ts/src/dataset/store.ts
/**
 * DatasetStore — JSON-file-backed CRUD for Dataset objects.
 *
 * Each dataset is stored as `<datasetDir>/<slug>.json`.
 * The directory is created on first write if it does not exist.
 */

import fs from "fs";
import path from "path";

import { getConfig } from "../config";
import type { Evaluable, JsonValue } from "../storage/evaluable";
import { UNSET } from "../storage/evaluable";
import type { Dataset } from "./models";

// ── Helpers ──────────────────────────────────────────────────────────────────

/**
 * Convert a dataset name to a filesystem-safe slug.
 *
 * Lowercase, replace non-alphanumeric runs with `-`, strip leading/trailing `-`.
 */
export function _slugify(name: string): string {
  const slug = name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
  if (!slug) {
    throw new Error(
      `Cannot slugify empty or non-alphanumeric name: ${JSON.stringify(name)}`
    );
  }
  return slug;
}

function timestampToIso(ts: number): string {
  const d = new Date(ts);
  // Format: YYYY-MM-DD HH:MM:SS
  return d.toISOString().replace("T", " ").replace(/\.\d{3}Z$/, "");
}

// ── Serialization helpers ────────────────────────────────────────────────────

function evaluableToJson(ev: Evaluable): Record<string, JsonValue> {
  const obj: Record<string, JsonValue> = {
    evalInput: ev.evalInput,
    evalOutput: ev.evalOutput,
    evalMetadata: ev.evalMetadata,
    expectedOutput:
      ev.expectedOutput === UNSET ? null : (ev.expectedOutput as JsonValue),
    evaluators: ev.evaluators as JsonValue,
    description: ev.description,
  };
  return obj;
}

function jsonToEvaluable(raw: Record<string, unknown>): Evaluable {
  return {
    evalInput: (raw["evalInput"] ?? raw["eval_input"] ?? null) as JsonValue,
    evalOutput: (raw["evalOutput"] ?? raw["eval_output"] ?? null) as JsonValue,
    evalMetadata:
      (raw["evalMetadata"] ??
        raw["eval_metadata"] ??
        null) as Record<string, JsonValue> | null,
    expectedOutput:
      raw["expectedOutput"] !== undefined
        ? (raw["expectedOutput"] as JsonValue)
        : raw["expected_output"] !== undefined
          ? (raw["expected_output"] as JsonValue)
          : UNSET,
    evaluators: (raw["evaluators"] as readonly string[] | null) ?? null,
    description: (raw["description"] as string | null) ?? null,
  };
}

// ── DatasetStore ─────────────────────────────────────────────────────────────

export class DatasetStore {
  private readonly _dir: string;

  constructor(datasetDir?: string) {
    this._dir = datasetDir ?? getConfig().datasetDir;
  }

  private _pathFor(name: string): string {
    return path.join(this._dir, `${_slugify(name)}.json`);
  }

  private _ensureDir(): void {
    fs.mkdirSync(this._dir, { recursive: true });
  }

  private _write(filePath: string, dataset: Dataset): void {
    this._ensureDir();
    const data = {
      name: dataset.name,
      items: dataset.items.map(evaluableToJson),
    };
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2) + "\n", "utf-8");
  }

  private _read(filePath: string): Dataset {
    const raw = JSON.parse(fs.readFileSync(filePath, "utf-8"));
    return {
      name: raw.name as string,
      items: (raw.items as Record<string, unknown>[]).map(jsonToEvaluable),
    };
  }

  // ── CRUD ──────────────────────────────────────────────────────────────

  /**
   * Create a new dataset.
   * @throws if a dataset with the same name already exists.
   */
  create(name: string, items?: Evaluable[]): Dataset {
    const filePath = this._pathFor(name);
    if (fs.existsSync(filePath)) {
      throw new Error(`Dataset already exists: ${JSON.stringify(name)}`);
    }
    const dataset: Dataset = { name, items: items ?? [] };
    this._write(filePath, dataset);
    return dataset;
  }

  /**
   * Load a dataset by name.
   * @throws if the dataset does not exist.
   */
  get(name: string): Dataset {
    const filePath = this._pathFor(name);
    if (!fs.existsSync(filePath)) {
      throw new Error(`Dataset not found: ${JSON.stringify(name)}`);
    }
    return this._read(filePath);
  }

  /** Return the names of all stored datasets. */
  list(): string[] {
    if (!fs.existsSync(this._dir)) return [];
    const names: string[] = [];
    const files = fs.readdirSync(this._dir).filter((f) => f.endsWith(".json")).sort();
    for (const f of files) {
      try {
        const ds = this._read(path.join(this._dir, f));
        names.push(ds.name);
      } catch {
        continue; // skip malformed files
      }
    }
    return names;
  }

  /** Return metadata for every stored dataset. */
  listDetails(): Array<Record<string, unknown>> {
    if (!fs.existsSync(this._dir)) return [];
    const rows: Array<Record<string, unknown>> = [];
    const files = fs.readdirSync(this._dir).filter((f) => f.endsWith(".json")).sort();
    for (const f of files) {
      try {
        const filePath = path.join(this._dir, f);
        const ds = this._read(filePath);
        const stat = fs.statSync(filePath);
        rows.push({
          name: ds.name,
          rowCount: ds.items.length,
          createdAt: timestampToIso(stat.birthtimeMs),
          updatedAt: timestampToIso(stat.mtimeMs),
        });
      } catch {
        continue; // skip malformed files
      }
    }
    return rows;
  }

  /**
   * Delete a dataset by name.
   * @throws if the dataset does not exist.
   */
  delete(name: string): void {
    const filePath = this._pathFor(name);
    if (!fs.existsSync(filePath)) {
      throw new Error(`Dataset not found: ${JSON.stringify(name)}`);
    }
    fs.unlinkSync(filePath);
  }

  /** Append items to an existing dataset. */
  append(name: string, ...items: Evaluable[]): Dataset {
    const dataset = this.get(name);
    const updated: Dataset = {
      name: dataset.name,
      items: [...dataset.items, ...items],
    };
    this._write(this._pathFor(name), updated);
    return updated;
  }

  /**
   * Remove an item by index from an existing dataset.
   * @throws if the dataset does not exist or index is out of range.
   */
  remove(name: string, index: number): Dataset {
    const dataset = this.get(name);
    const items = [...dataset.items];
    if (index < 0 || index >= items.length) {
      throw new RangeError(
        `Index ${index} out of range for dataset ${JSON.stringify(name)} with ${items.length} items`
      );
    }
    items.splice(index, 1);
    const updated: Dataset = { name: dataset.name, items };
    this._write(this._pathFor(name), updated);
    return updated;
  }
}
